### PR TITLE
Improve null and object printing to avoid confusion with arrays

### LINKED ISSUE #4944: Changing `[Reference]` to `<Reference>` or `(Reference)` for printing
### Describe the project you are working on

A simple dice game where I record each action as a list of Action classes

### Describe the problem or limitation you are having in your project

When you print an array of references, this comes up:
```
[[Reference:2937], [Reference:2938]]
```
This makes it look like an array of 2 arrays containing references, but it's an array of 2 reference, because if we print reference by itself it becomes `[Reference:2937]`.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Changing to `<Reference:2937>` or `(Reference:2937)` would remove this confusion.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

It will fix this in default functionality.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You can write your own formatting/printing functions.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think there's a particular reason for using [] for formatting references, so using <> or () would fix this issue.

### PR DESCRIPTION
**Pull request by Calinou**
- Use different syntax for object printing to avoid confusion with arrays.
- ~~Add a leading and trailing space within directory printing to avoid confusion with arrays on some fonts.~~
  - Split into another PR: https://github.com/godotengine/godot/pull/64577
- Print null as `<null>` to avoid confusion with a string `"null"`.
- Display `<empty>` in editor resource pickers to avoid confusion with array-based properties.

The editor parts of this PR may be redoable in a different PR for 3.6. I'm not sure if the rest can be changed without potentially breaking compatibility though.

This closes https://github.com/godotengine/godot-proposals/issues/4944.

**Testing project:** [test_reference_print.zip](https://github.com/godotengine/godot/files/9176824/test_reference_print.zip)

## Preview

Before | After
-|-
 | 

### Before

```
null
null

[Curve:-9223372012544589442]
[Curve:-9223372012527812225]

Node2D:[Node2D:24226300274]
Node:[Node:24243077655]

[]
[0, 1, 2, 3]
{}
{"hello":"world", "one":"two"}
```

### After

```
<null>
null

<Curve#-9223372012393594492>
<Curve#-9223372012376817278>

Node2D:<Node2D#24377295223>
Node:<Node#24394072451>

[]
[0, 1, 2, 3]
{  }
{ "hello": "world", "one": "two" }
```

## TODO

- [ ] Update unit tests before merging.

### CODE REVIEW

File: core/variant/variant.cpp [hunk 1]
```diff
@@ -1700,11 +1700,13 @@ String Variant::stringify(int recursion_count) const {
...
-				ERR_PRINT("Max recursion reached");
-				return "{...}";
+				ERR_PRINT("Maximum dictionary recursion reached!");
+				return "{ ... }";
 			}
 
-			String str("{");
+			// Add leading and trailing space to Dictionary printing. This distinguishes it
+			// from array printing on fonts that have similar-looking {} and [] characters.
+			String str("{ ");
```
Comment (nathanfranke) [inline core/variant/variant.cpp:1709]: Only change I'm not 100% about. Leads to an "asymmetry" of sorts.

**Review by nathanfranke — APPROVED** [review]

File: core/variant/variant.cpp (hunk 1 above)
Comment (Calinou) [reply core/variant/variant.cpp:1709]: It's common for JavaScript formatters to format single-line arrays like this `[1, 2]`, but objects like this `{ hello: "world" }`.

Comment (nathanfranke) [reply core/variant/variant.cpp:1709]: I see what you mean! 
javascript console logging two values, an empty object and an empty array. the object shows spaces between the curly braces

**Comment by fire-forge** [comment]
Overall most of these changes are good, except for the extra spaces that are added at the beginning and end of dictionaries.

This example:
```
{  }
{ "hello": "world", "one": "two" }
```

Would be better as this:
```
{ }
{"hello": "world", "one": "two"}
```

File: core/variant/variant.cpp (hunk 1 above)
Comment (nathanfranke) [inline core/variant/variant.cpp:1709]: This would work:
```suggestion
			String str("{");
```

**Review by nathanfranke — COMMENTED** [review]
> Would be better as this:
> 
> ```
> { }
> {"hello": "world", "one": "two"}
> ```

~~I'm ready for the bikeshedding :laughing: (which space do we remove)~~

**Comment by fire-forge** [comment]
> I'm ready for the bikeshedding 😆 (which space do we remove)

I actually meant to remove both spaces, so like this instead:

```
{}
{"hello": "world", "one": "two"}
```

**Comment by Calinou** [comment]
> Discussed in PR review meeting. We agreed that this PR should only cover the things requested and discussed in the proposal i.e. using angle brackets to distinguish objects from arrays.

Done. I split the Dictionary changes to another [PR](https://github.com/godotengine/godot/pull/64577) and [proposal](https://github.com/godotengine/godot-proposals/issues/5198).

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Needs rebase. And GDScript unit tests need to be fixed: https://github.com/godotengine/godot/runs/7901811603?check_suite_focus=true

**Pull request merged**